aiida/manage/external/rmq.py

```diff
@@ -16,48 +16,57 @@
 import plumpy
 from kiwipy import communications, Future
 
-__all__ = ('RemoteException', 'CommunicationTimeout', 'DeliveryFailed', 'ProcessLauncher')
+from aiida.common.extendeddicts import AttributeDict
+
+__all__ = ('RemoteException', 'CommunicationTimeout', 'DeliveryFailed', 'ProcessLauncher', 'BROKER_DEFAULTS')
 
 LOGGER = logging.getLogger(__name__)
 
 RemoteException = plumpy.RemoteException
 DeliveryFailed = plumpy.DeliveryFailed
 CommunicationTimeout = communications.TimeoutError  # pylint: disable=invalid-name
 
-# GP: Using here 127.0.0.1 instead of localhost because on some computers
-# localhost resolves first to IPv6 with address ::1 and if RMQ is not
-# running on IPv6 one gets an annoying warning. When moving this to
-# a user-configurable variable, make sure users are aware of this and
-# know how to avoid warnings. For more info see
-# https://github.com/aiidateam/aiida-core/issues/1142
-_RMQ_URL = 'amqp://127.0.0.1'
-_RMQ_HEARTBEAT_TIMEOUT = 600  # Maximum that can be set by client, with default RabbitMQ server configuration
 _LAUNCH_QUEUE = 'process.queue'
 _MESSAGE_EXCHANGE = 'messages'
 _TASK_EXCHANGE = 'tasks'
-
-
-def get_rmq_url(heartbeat_timeout=None):
-    """
-    Get the URL to connect to RabbitMQ
-
-    :param heartbeat_timeout: the interval in seconds for the heartbeat timeout
+BROKER_DEFAULTS = AttributeDict({
+    'protocol': 'amqp',
+    'username': 'guest',
+    'password': 'guest',
+    'host': '127.0.0.1',
+    'port': 5672,
+})
+
+
+def get_rmq_url(protocol=None, username=None, password=None, host=None, port=None, heartbeat_timeout=600):
+    """Return the URL to connect to RabbitMQ.
+
+    .. note::
+
+        The default of the ``host`` is set to ``127.0.0.1`` instead of ``localhost`` because on some computers localhost
+        resolves first to IPv6 with address ::1 and if RMQ is not running on IPv6 one gets an annoying warning. For more
+        info see: https://github.com/aiidateam/aiida-core/issues/1142
+
+    :param username: the username for authentication
+    :param password: the password for authentication
+    :param host: the hostname of the RabbitMQ server
+    :param port: the port of the RabbitMQ server
+    :param heartbeat_timeout: the interval in seconds for the heartbeat timeout, default of 600 is maximum that can be
+        set by client, with default RabbitMQ server configuration.
     :returns: the connection URL string
     """
-    url = _RMQ_URL
-
-    if heartbeat_timeout is None:
-        heartbeat_timeout = _RMQ_HEARTBEAT_TIMEOUT
-
-    if heartbeat_timeout is not None:
-        url += '?heartbeat={}'.format(heartbeat_timeout)
-
-    return url
+    return '{protocol}://{username}:{password}@{host}:{port}?heartbeat={heartbeat_timeout}'.format(
```

Comment:
You might want to consider using [urlparse](https://docs.python.org/3.5/library/urllib.parse.html?highlight=urlparse#urllib.parse.urlunparse) for this.

Reply:
Changed it to use `urlparse`. Should note though that the `{username}:{password}@{host}:{port}` part is not formatted by it though. It understands that entire part as the `netloc` and so it us up to us to format it. So there I still use string interpolation as before